### ISSUE TITLE
Changes made by Rancher outside of Pulumi are not detected by Pulumi

### ISSUE THREAD

**Issue by ghost** [pulumi/pulumi-kubernetes#641]
Creating a service like so:
```
new k8s.core.v1.Service(appName, {
    metadata: {
        namespace: 'sia-ns',
        name: 'nginx-service'
    },
    spec: {
        type: 'ClusterIP',
        ports: [{ port: 80, targetPort: 8080, protocol: 'TCP'}],
        selector: {
            app: 'nginx'
        }
    }
});
```
after updating the service over another channel to `targetPort: 8081`

pulumi doesn't detect the change. Refresh appears to sometimes notice, but following updates don't restore the targetPort to 8080

**Comment by lukehoban** [comment]
> Refresh appears to sometimes notice,

Do you mean it sometimes works but other times doesn't?  That would be very unexpected!

> following updates don't restore the targetPort to 8080

That is definitely a bug.

**Comment by pgavlin** [comment]
What versions of the Pulumi CLI and `@pulumi/Kubernetes` are you using?

**Comment by aeb-sia** [comment]
I'm using pulumi v0.17.22 and @pulumi/kubernetes 0.25.2

**Comment by aeb-sia** [comment]
I did some more testing.
It looks like if I edit the service using kubectl
```
apiVersion: v1
kind: Service
metadata:
  namespace: sia-ns
  name: nginx-service
spec:
  selector: 
    App: nginx
  type: ClusterIP
  ports:
    - name: http
      port: 80
      targetPort: 8081
      protocol: TCP
```
Pulumi refresh notices the change and subsequent updates work as expected.

If however I edit the service using rancher, pulumi refresh still picks up the change, but doesn't modify the stack's state in such a way that subsequent updates overwrite it.

Describe after pulumi:
```
 % kubectl describe service/nginx-service
Name:              nginx-service
Namespace:         sia-ns
Labels:            app.kubernetes.io/managed-by=pulumi
Annotations:       kubectl.kubernetes.io/last-applied-configuration:
                     {"apiVersion":"v1","kind":"Service","metadata":{"labels":{"app.kubernetes.io/managed-by":"pulumi"},"name":"nginx-service","namespace":"sia...
Selector:          App=nginx
Type:              ClusterIP
IP:                100.78.175.37
Port:              http  80/TCP
TargetPort:        8080/TCP
Endpoints:         100.80.150.23:8080
Session Affinity:  None
Events:            <none>
```
After kubectl update
```
 % kubectl describe service/nginx-service
Name:              nginx-service
Namespace:         sia-ns
Labels:            <none>
Annotations:       kubectl.kubernetes.io/last-applied-configuration:
                     {"apiVersion":"v1","kind":"Service","metadata":{"annotations":{},"name":"nginx-service","namespace":"sia-ns"},"spec":{"ports":[{"name":"ht...
Selector:          App=nginx
Type:              ClusterIP
IP:                100.78.175.37
Port:              http  80/TCP
TargetPort:        8081/TCP
Endpoints:         100.80.150.23:8081
Session Affinity:  None
Events:            <none>
```
After edit in rancher
```
 % kubectl describe service/nginx-service
Name:              nginx-service
Namespace:         sia-ns
Labels:            app.kubernetes.io/managed-by=pulumi
Annotations:       field.cattle.io/ipAddresses: null
                   field.cattle.io/targetDnsRecordIds: null
                   field.cattle.io/targetWorkloadIds: null
                   kubectl.kubernetes.io/last-applied-configuration:
                     {"apiVersion":"v1","kind":"Service","metadata":{"labels":{"app.kubernetes.io/managed-by":"pulumi"},"name":"nginx-service","namespace":"sia...
Selector:          App=nginx
Type:              ClusterIP
IP:                100.76.108.249
Port:              http  80/TCP
TargetPort:        8081/TCP
Endpoints:         100.80.150.23:8081
Session Affinity:  None
Events:            <none>
```

(The IP change is because I delete the service and redeploy it using pulumi to start with a clean service)

**Comment by lukehoban** [comment]
Thanks @aeb-sia.  I've updated the title to match the details of your investigation.

cc @hausdorff for thoughts on this as well.

**Comment by pgavlin** [comment]
This is a bit surprising... I would have expected that the three-way-merge logic used in `Diff` would pick this up.

As an aside, I'm glad that we've found a tool that does not update that annotation, as it should make these scenarios much easier to test.

**Comment by hausdorff** [comment]
Talking to @lukehoban again, let's move this back provisionally to Q3. I'll take some time in the next couple days to dig in a bit at the intersection of diff and refresh to build more confidence that this works as expected.

@aeb-sia in the mean time, do you have a list of specific steps you took to arrive in this situation? It would be very nice to have a repro.

**Comment by aeb-sia** [comment]
@hausdorff create the service as described above then change the port in rancher.

**Comment by lblackstone** [comment]
Haven't had time to look into this further, so bumping to M26

**Comment by lblackstone** [comment]
Reproduced the underlying issue in #694. This does work properly with the dryRun-based diff behavior, but it was causing some other stability problems, so it's disabled by default. If you want to try it out, you can set the `enableDryRun` flag on the provider for the resource. e.g.,

```ts
const provider = new k8s.Provider("foo", {enableDryRun: true});
```

**Comment by lblackstone** [comment]
This should be fixed with the `enableServerSideApply` mode available in the [v3.20.1 release](https://github.com/pulumi/pulumi-kubernetes/releases/tag/v3.20.1)
